sdl_viewer/src/box_drawer.rs

```diff
@@ -0,0 +1,144 @@
+
+pub struct BoxDrawer {
+    outline_program: GlProgram,
+
+    // Uniforms locations.
+    u_transform: GLint,
+    u_color: GLint,
+
+    // Vertex array and buffers
+    vertex_array: GlVertexArray,
+    _buffer_position: GlBuffer,
+    _buffer_indices: GlBuffer,
+}
+
+impl BoxDrawer {
+    pub fn new() -> Self {
+        let outline_program =
+            GlProgram::new(VERTEX_SHADER_OUTLINED_BOX, FRAGMENT_SHADER_OUTLINED_BOX);
+        let u_transform;
+        let u_color;
+
+        unsafe {
+            gl::UseProgram(outline_program.id);
+            u_transform = gl::GetUniformLocation(outline_program.id, c_str!("transform"));
+            u_color = gl::GetUniformLocation(outline_program.id, c_str!("color"));
+        }
+
+        let vertex_array = GlVertexArray::new();
+        vertex_array.bind();
+
+        // vertex buffer: define 8 vertices of the box
+        let _buffer_position = GlBuffer::new_array_buffer();
+        _buffer_position.bind();
+        let vertices: [[f32; 3]; 8] = [
+            [-1.0, -1.0,  1.0],   // vertices of front quad
+            [ 1.0, -1.0,  1.0],   //
+            [ 1.0,  1.0,  1.0],   //
+            [-1.0,  1.0,  1.0],   //
+            [-1.0, -1.0, -1.0],   // vertice ofs back quad
+            [ 1.0, -1.0, -1.0],   //
+            [ 1.0,  1.0, -1.0],   //
+            [-1.0,  1.0, -1.0],   //
+        ];
+        unsafe {
+            gl::BufferData(
+                gl::ARRAY_BUFFER,
+                (vertices.len() * 3 * mem::size_of::<f32>()) as GLsizeiptr,
+                mem::transmute(&vertices[0]),
+                gl::STATIC_DRAW,
+            );
+        }
+
+        // define index buffer for 24 edges of the box
+        let _buffer_indices = GlBuffer::new_element_array_buffer();
+        _buffer_indices.bind();
+        let indices: [i32; 24] = [
+            0,1, 1,2, 2,3, 3,0,		// front
+            4,5, 5,6, 6,7, 7,4,		// back
+            1,5, 6,2,				// right
+            4,0, 3,7,				// left
+        ];
+        unsafe {
+            gl::BufferData(
+                gl::ELEMENT_ARRAY_BUFFER,
+                (indices.len() * mem::size_of::<i32>()) as GLsizeiptr,
+                mem::transmute(&indices[0]),
+                gl::STATIC_DRAW,
+            );
+        }
+
+        unsafe {
+            let pos_attr = gl::GetAttribLocation(outline_program.id, c_str!("position"));
+            gl::EnableVertexAttribArray(pos_attr as GLuint);
+            gl::VertexAttribPointer(
+                pos_attr as GLuint,
+                3,
+                gl::FLOAT,
+                gl::FALSE,
+                3 * mem::size_of::<f32>() as i32,
+                ptr::null(),
+            );
+        }
+        BoxDrawer {
+            outline_program,
+            u_transform,
+            u_color,
+            vertex_array,
+            _buffer_position,
+            _buffer_indices,
+        }
+    }
+
+    // draws the unit box using any kind of transformation
```

Comment:
This is unclear to me. You explained me on the whiteboard, but please add a clarifying sentence here too. Something like: Draws the outline of the box where each vertex is transformed with `matrix`.

Reply:
Also `matrix` is the type. Rename to `transform`?